pkg/tcpip/stack/ndp.go

```diff
@@ -720,6 +743,83 @@ func (ndp *ndpState) handleRA(ip tcpip.Address, ra header.NDPRouterAdvert) {
+	// remote link address can be calculated without a resolution process.
+	if c, err := r.Resolve(nil); err != nil {
+		log.Fatalf("ndp: error when resolving route to send NDP RA (%s -> %s on NIC(%d)): %s", header.IPv6Any, header.IPv6AllRoutersMulticastAddress, ndp.nic.ID(), err)
+	} else if c != nil {
+		log.Fatalf("ndp: route resolution not immediate for route to send NDP RA (%s -> %s on NIC(%d))", header.IPv6Any, header.IPv6AllRoutersMulticastAddress, ndp.nic.ID())
+	}
+
+	pi := header.NewNDPPrefixInformation()
+	pi.SetPrefix(ndp.configs.routerConfig.Prefix)
+	pi.SetPrefixLength(ndp.configs.routerConfig.PrefixLength)
+	pi.SetOnLinkFlag(ndp.configs.routerConfig.OnLinkFlag)
+	pi.SetAutonomousAddressConfigurationFlag(ndp.configs.routerConfig.AutonomousAddressConfigurationFlag)
+	pi.SetPreferredLifetime(ndp.configs.routerConfig.PreferredLifetime)
+	pi.SetValidLifetime(ndp.configs.routerConfig.ValidLifetime)
+
+	optsSerializer := header.NDPOptionsSerializer{
+		header.NDPSourceLinkLayerAddressOption(r.LocalLinkAddress[:]),
+		pi,
+	}
+	hdr := buffer.NewPrependable(int(r.MaxHeaderLength()) + header.ICMPv6RouterAdvertMinimumSize + int(optsSerializer.Length()))
+
+	packet := header.ICMPv6(hdr.Prepend(header.ICMPv6RouterAdvertSize))
+	packet.SetType(header.ICMPv6RouterAdvert)
+
+	ra := header.NDPRouterAdvert(packet.NDPPayload())
+	ra.SetRouterLifetime(ndp.configs.routerConfig.RouterLifetime)
+	ra.SetCurrHopLimit(ndp.configs.routerConfig.CurrHopLimit)
+	opts := ra.Options()
+	opts.Serialize(optsSerializer)
+
+	// r.LocalAddress = tcpip.Address(net.ParseIP("fe80::2"))
+	// r.RemoteAddress = header.IPv6AllNodesMulticastAddress
+	// r.RemoteLinkAddress = tcpip.LinkAddress([]byte{0x33, 0x33, 0x00, 0x00, 0x00, 0x01})
+
+	packet.SetChecksum(header.ICMPv6Checksum(packet, r.LocalAddress, r.RemoteAddress, buffer.VectorisedView{}))
+
+	sent := r.Stats().ICMP.V6PacketsSent
+	if err := r.WritePacket(nil,
+		NetworkHeaderParams{
+			Protocol: header.ICMPv6ProtocolNumber,
+			TTL:      header.NDPHopLimit,
+			TOS:      DefaultTOS,
+		}, tcpip.PacketBuffer{Header: hdr},
+	); err != nil {
+		sent.Dropped.Increment()
+		log.Printf("startSolicitingRouters: error writing NDP router solicit message on NIC(%d); err = %s", ndp.nic.ID(), err)
+	} else {
+		sent.RouterSolicit.Increment()
+	}
+}
+
+// handleRS handles a Router Solicitation message that arrived on the NIC
+// this ndp is for. Does nothing if the NIC is configured to not handle RSs.
+//
+// The NIC that ndp belongs to MUST be locked.
+func (ndp *ndpState) handleRS(rs header.NDPRouterSolicit) {
+	// Is the NIC configured to handle RSs at all?
+	//
+	// Currently, the stack does not determine router interface status on a
+	// per-interface basis; it is a stack-wide configuration, so we check
+	// stack's forwarding flag to determine if the NIC is a routing
+	// interface.
+	if !ndp.configs.HandleRSs || !ndp.nic.stack.forwarding {
+		return
+	}
+
+	go ndp.sendRouterAdvert()
```

Comment:
There is some rate limiting requirements we need to follow for outgoing RAs. Please see RFC 4861 section 6.2.6 https://tools.ietf.org/html/rfc4861#section-6.2.6

Reply:
I have done basic rate limiting, I left some TODO because it should also handle unsollicited RAs, and probably use x/time/rate.